src/services/grpc/api/v2/login.ts

```diff
@@ -0,0 +1,89 @@
+import { Status, ServerError } from 'nice-grpc';
+import bcrypt from 'bcrypt';
+import { getPNIDByUsername, getPNIDByTokenAuth } from '@/database';
+import { nintendoPasswordHash, generateToken } from '@/util';
+import { config } from '@/config-manager';
+import type { LoginRequest, LoginResponse, DeepPartial } from '@pretendonetwork/grpc/api/v2/login_rpc';
+import type { HydratedPNIDDocument } from '@/types/mongoose/pnid';
+
+export async function login(request: LoginRequest): Promise<DeepPartial<LoginResponse>> {
+	const grantType = request.grantType?.trim();
+	const username = request.username?.trim();
+	const password = request.password?.trim();
+	const refreshToken = request.refreshToken?.trim();
+
+	if (!['password', 'refresh_token'].includes(grantType)) {
+		throw new ServerError(Status.INVALID_ARGUMENT, 'Invalid grant type');
+	}
+
+	if (grantType === 'password' && !username) {
+		throw new ServerError(Status.INVALID_ARGUMENT, 'Invalid or missing username');
+	}
+
+	if (grantType === 'password' && !password) {
+		throw new ServerError(Status.INVALID_ARGUMENT, 'Invalid or missing password');
+	}
+
+	if (grantType === 'refresh_token' && !refreshToken) {
+		throw new ServerError(Status.INVALID_ARGUMENT, 'Invalid or missing refresh token');
+	}
+
+	let pnid: HydratedPNIDDocument | null;
+
+	if (grantType === 'password') {
+		pnid = await getPNIDByUsername(username!); // * We know username will never be null here
+
+		if (!pnid) {
+			throw new ServerError(Status.INVALID_ARGUMENT, 'User not found');
+		}
+
+		const hashedPassword = nintendoPasswordHash(password!, pnid.pid); // * We know password will never be null here
+
+		if (!bcrypt.compareSync(hashedPassword, pnid.password)) {
+			throw new ServerError(Status.INVALID_ARGUMENT, 'Password is incorrect');
+		}
+	} else {
+		pnid = await getPNIDByTokenAuth(refreshToken!); // * We know refreshToken will never be null here
+
+		if (!pnid) {
+			throw new ServerError(Status.INVALID_ARGUMENT, 'Invalid or missing refresh token');
+		}
+	}
+
+	if (pnid.deleted) {
+		throw new ServerError(Status.UNAUTHENTICATED, 'Account has been deleted');
+	}
+
+	const accessTokenOptions = {
+		system_type: 0x3, // * API
+		token_type: 0x1, // * OAuth Access
+		pid: pnid.pid,
+		access_level: pnid.access_level,
+		title_id: BigInt(0),
+		expire_time: BigInt(Date.now() + (3600 * 1000))
+	};
+
+	const refreshTokenOptions = {
+		system_type: 0x3, // * API
+		token_type: 0x2, // * OAuth Refresh
+		pid: pnid.pid,
+		access_level: pnid.access_level,
+		title_id: BigInt(0),
+		expire_time: BigInt(Date.now() + (3600 * 1000))
```

Comment:
While you're here, would you like to increase the refresh token expiry to be > access token?

Reply:
I planned to update token times in a separate PR, the one Quarky and I talked about on Discord the other day with regards to service tokens

Reply:
Cool 👍